app/controllers/concerns/hud_reports/artifact_aware_cells.rb

```diff
@@ -0,0 +1,135 @@
+###
+# Copyright 2016 - 2025 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+# frozen_string_literal: true
+
+# Concern for handling client loading in cells controllers when artifacts are stored in S3
+module HudReports
+  module ArtifactAwareCells
+    require_relative 'temporary_cell_row'
+    require_relative 'temporary_cell_collection'
+    extend ActiveSupport::Concern
+
+    private
+
+    # Load clients for a specific cell, handling both RDS and S3 data sources
+    def load_cell_clients(client_class, cell, table)
+      if @report.artifacts_stored?
+        load_clients_from_s3(client_class, cell, table)
+      else
+        load_clients_from_rds(client_class, cell, table)
+      end
+    end
+
+    # Load clients using legacy database joins.
+    # This will only return clients for reports whose data HAS NOT been stored in S3.
+    # When a report is stored in S3, the data in these tables are cleared out.
+    def load_clients_from_rds(client_class, cell, table)
+      client_class.
+        joins(hud_reports_universe_members: { report_cell: :report_instance }).
+        merge(::HudReports::ReportCell.for_table(table).for_cell(cell)).
+        merge(::HudReports::ReportInstance.where(id: @report.id))
+    end
+
+    # Load clients using S3 data.
+    # This will only return clients for reports whose data HAS been stored in S3.
+    def load_clients_from_s3(client_class, cell, table)
```

Comment:
Let's rename this method too:
```suggestion
    def load_clients_from_file(client_class, cell, table)
```

Reply:
Included in [this commit](https://github.com/greenriver/hmis-warehouse/pull/5644/commits/189c4cdec097138b0dbd1723976918957ed2e9d8)